eip-5516 draft (#5516)
ethereum/EIPs: EIPS/eip-5516.md

---
eip: 5516
title: Soulbound, Multi Token Standard.
description: An interface for non-transferrable NFTs binding to an Ethereum account for standardized blockchain-based academic certification.

> author: Lucas Martín Grasso Ramos(@LucasGrasso), Matias Arazi(@MatiArazi)

Reviewer: ```suggestion
author: Lucas Martín Grasso Ramos (@LucasGrasso), Matias Arazi (@MatiArazi)
```
Required: Formatting

discussions-to: https://ethereum-magicians.org/t/eip-5516-soulbound-multi-token-standard/10485
status: Draft
type: Standards Track
category: ERC
created: 2022-08-19
requires: 165, 1155
---

## Abstract

> This is a standard interface for non fungible, double signature, multi-token, Soulbound Token standard. It was inspired by the main characteristics of the EIP-1155 token and by the [article](https://vitalik.ca/general/2022/01/26/soulbound.html) and [paper](https://papers.ssrn.com/sol3/papers.cfm?abstract_id=4105763) published by Vitalik Butherin, in which benefits and potential use-cases of SoulBoundTokens(SBT) were presented.

Reviewer: Required: No external links

## Motivation

We found a potential problem, the lack of credibility and faithfulness in online certifications, and imagined a solution, inspired by the everyday more common articles and mentions of Soulbound Tokens. We set to combine the best characteristics of each standardized EIP with the sole objective of elaborating a solid and useful token standard.

### Characteristics
* The NFT will be nontransferable after the initial transfer(SoulBoundToken-SBT).
* Is backwards compatible with EIP-1155.
* Double Signature.
* Multi-Token.
* Semi-Fungible standard.

### Applications
* Academy
* Certifications
* Smart Contract certifications(Code auditories)
* POAP
* And more

## Specification
The key words “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL NOT”, “SHOULD”, “SHOULD NOT”, “RECOMMENDED”, “MAY”, and “OPTIONAL” in this document are to be interpreted as described in RFC 2119.

> **Smart contracts implementing the SBTERC1155DS standard MUST implement all of the functions in the SBTERC1155DS interface.**

Reviewer: ```suggestion
**Smart contracts implementing this EIP MUST implement all of the functions in the `SBTERC1155DS` interface.**
```
Required: Self-reference as "this EIP"

> **Smart contracts implementing the SBTERC1155DS standard MUST implement the ERC-165 supportsInterface function and and MUST return the constant value true if 0x812b9fae is passed through the interfaceID argument.They also MUST implement the EIP-1155 Interface and MUST return the constant value true if 0xd9b67a26 is passed through the interfaceID argument. Furthermore, they MUST implement the EIP-1155 Metadata interface, and MUST return the constant value true if 0x0e89341c is passed through the interfaceID argument.**

Reviewer: ```suggestion
**Smart contracts implementing this EIP MUST implement the [EIP-165](./eip-165.md) `supportsInterface` function and and MUST return the constant value true if `0x812b9fae` is passed through the `interfaceID` argument. They also MUST implement the EIP-1155 Interface and MUST return the constant value `true` if `0xd9b67a26` is passed through the `interfaceID` argument. Furthermore, they MUST implement the EIP-1155 Metadata interface, and MUST return the constant value true if `0x0e89341c` is passed through the `interfaceID` argument.**
```
Required: Self-reference as "this EIP"
Required: Grammar
Required: Reference as `EIP-<number>`
Required: Link on the first mention of an EIP
Recommended: Formatting

> See [EIP-1155 Specification](https://eips.ethereum.org/EIPS/eip-1155#specification)

Reviewer: ```suggestion
See [EIP-1155](./eip-1155.md#specification)
```
Required: Internally link where possible

```solidity

> // SPDX-License-Identifier: MIT

Reviewer 2: [editorial] Please update license to CC0-1.0 here an everywhere else

Reviewer 3: Please update license to CC0-1.0 here an everywhere else

    /**
     * @dev Get tokens owned by a given address
     */
    function tokensFrom(address from) external view returns (uint256[] memory);

    /**
     * @dev Get tokens marked as pending of a given address
     */
    function pendingFrom(address from) external view returns (uint256[] memory);
    
    /**
     * @dev Claims or Reject pending `_id` from address `_account`.
     *
     * Requirements:
     * - `account` cannot be the zero address.
     * - `account` MUST have a pending token under `id` at the moment of call.
     * - `account` MUST not own a token under `id` at the moment of call.
     *
     * Emits a {TokenClaimed} event.
     *
     */
    function claimOrReject(uint256 _id,bool _action) external;

    /**
     * @dev Transfers `_id` token from `_from` to every address at `_to[]`.
     *
     * Requirements:
     *
     * - `_from` MUST be the creator(minter) of `id`.
     * - All addresses in `to[]` MUST be non-zero.
     * - All addresses in `to[]` MUST have the token `id` under `_pendings`.
     * - All addresses in `to[]` MUST not own a token type under `id`.
     *
     * Emits a {TransfersMulti} event.
     *
     */

>     function batchTransfer (address from, address[] memory to, uint256 id, bytes memory data) external;

Reviewer 2: This seems not compatible with EIP-1155, can you double check?

### Only transferable by token creator/minter or allowed operators (Reference Implementation)
As the problem this standard aims to solve is academic certification, this characteristic also tends to mimic real world functionality. Only the creator/minter of the token under `id` will be able to transfer the token to new people (Semi-Fungible token). This allows for issuer traceability and connections with a DAO Voting System, for example, in order to only allow emission to voted and trusted addresses. Regardless of this, users have full power over their tokens: They can claim/reject them, and burn them afterwards. Token creators/minters have no control over a token after minting.

### Custom error messages instead of require statements.
After testing the standard, we found out that using custom errors was way more gas efficient than using require()
 statements.
### Metadata.
The EIP1155 Metadata Interface was implemented for more compatibility with EIP-1155.

> See [EIP-1155 Metadata](https://eips.ethereum.org/EIPS/eip-1155#metadata)

Reviewer: ```suggestion
See [EIP-1155](./eip-1155.md#metadata)
```
Required: Reference EIPS as `EIP-<number>`
Required: Internally link where possible

### Multi token
The multi token functionality was used in order for easier organization, all emitted tokens are stored in the same contract, preventing redundant bytecode from being deployed to the blockchain. It also facilitates transfer to token issuers, given that all issued tokens are stored and can be accessed under the same address.

### batchTransfer() function
The token standard supports a batch transfer function to multiple addresses, and this was made to ease the transfer to multiple people, making it more gas-efficient and easier for the users.

## Backwards Compatibility
This proposal is fully backward compatible with EIP-1155.

## Security Considerations
There are no security considerations related directly to the implementation of this standard.

> ## References
> 
> **Standards**
> * [EIP-1155 Multi Token Standard](https://eips.ethereum.org/EIPS/eip-1155)
> * [EIP-165 Standard Interface Detection](https://eips.ethereum.org/EIPS/eip-165)
> * [JSON Schema](https://json-schema.org/)
> * [RFC 2119 Key words for use in RFCs to Indicate Requirement Levels](https://www.ietf.org/rfc/rfc2119.txt)
> 
> **Implementations**
> - [SBTERC1155DS Reference Implementation](https://ipfs.io/ipfs/QmWyU2KxpNzWCm5tPWgm517KFvaTtY1aehNDM1fJXL3eEd?filename=SBTERC1155DS.sol)
> * [Zerti](https://dev.zerti.com.ar/feed) (Link to Rinkeby testnet test application)
> 
> **Articles & Discussions**
> * [Decentralized Society: Finding Web3's Soul](https://papers.ssrn.com/sol3/papers.cfm?abstract_id=4105763)
> * [Soulbound](https://vitalik.ca/general/2022/01/26/soulbound.html)
> 

Reviewer: ```suggestion
```
Required: Invalid section
Required: No external links

---
eip: 5516

> title: Soulbound, Multi-owned token.

Reviewer: ```suggestion
title: Soulbound Multi-owned tokens
```

description: An interface for non-transferrable, Multi-owned NFTs binding to Ethereum accounts.
author: Lucas Martín Grasso Ramos (@LucasGrasso), Matias Arazi (@MatiArazi)
discussions-to: https://ethereum-magicians.org/t/EIP-5516-soulbound-multi-token-standard/10485
status: Draft
type: Standards Track
category: ERC
created: 2022-08-19
requires: 165, 1155
---

## Abstract

> Proposes a standard interface for non-fungible, double signature, multi-token, Soulbound Tokens. It was inspired by the main characteristics of the [EIP-1155](./eip-1155.md) token and by articles in which benefits and potential use-cases of Soulbound/Accountbound Tokens(SBTs) were presented.

Reviewer: ```suggestion
Proposes a standard interface for non-fungible, double signature, multi-token, Soulbound Tokens. It was inspired by the main characteristics of the [EIP-1155](./eip-1155.md) token and by articles in which benefits and potential use-cases of Soulbound/Accountbound Tokens (SBTs) were presented.
```

> Proposes a standard interface for non-fungible, double signature, multi-token, Soulbound Tokens. It was inspired by the main characteristics of the [EIP-1155](./eip-1155.md) token and by articles in which benefits and potential use-cases of Soulbound/Accountbound Tokens (SBTs) were presented.

Reviewer: ```suggestion
This EIP proposes a standard interface for non-fungible double signature Soulbound multi-tokens. Previous account-bound token standards face the issue of users losing their account keys or having them rotated, thereby losing their tokens in the process. This EIP provides a solution to this issue and allows for the recycling of SBTs: minting of NFTs with the same characteristics for it being transferred to different addresses is no longer necessary.
```
Recommended: Moved parts of the Motivation to the Abstract
Recommended: Rephrasing

## Motivation

> Soulbound/Accountbound token standards like [EIP-4973](./eip-4973.md) face the issue of users losing their account keys or having them rotated, thereby losing the binding to SBTs. This [EIP-1155](./eip-1155.md)-inspired interface design provides a solution to this issue and allows for the recycling of SBTs: minting of NFTs with the same characteristics for it being transferred to different addresses is no longer necessary.

Reviewer: ```suggestion
This EIP was inspired by the main characteristics of the [EIP-1155](./eip-1155.md) token and by articles in which benefits and potential use cases of Soulbound/Accountbound Tokens (SBTs) were presented. 
```
Recommended: Moved this to the Abstract, put some stuff from the Abstract here

> Soulbound/Accountbound token standards like [EIP-4973](./eip-4973.md) face the issue of users losing their account keys or having them rotated, thereby losing the binding to SBTs. This [EIP-1155](./eip-1155.md)-inspired interface design provides a solution to this issue and allows for the recycling of SBTs: minting of NFTs with the same characteristics for it being transferred to different addresses is no longer necessary.  

Reviewer: ```suggestion
Soulbound/Accountbound token standards like [EIP-4973](./eip-4973.md) face the issue of users losing their account keys or having them rotated, thereby losing the binding to SBTs. This [EIP-1155](./eip-1155.md)-inspired interface design provides a solution to this issue and allows for the recycling of SBTs: minting of NFTs with the same characteristics for it being transferred to different addresses is no longer necessary.
```

> This design also allows for multiple token types to be transferred at once, saving on transaction costs. Trading of multiple tokens can be built on top of this standard and it removes the need to approve individual token contracts separately. It is also easy to describe and mix multiple fungible or non-fungible token types in a single contract.

Reviewer: ```suggestion
This design also allows for batch token transfers, saving on transaction costs. Trading of multiple tokens can be built on top of this standard and it removes the need to approve individual token contracts separately. It is also easy to describe and mix multiple fungible or non-fungible token types in a single contract.
```
Recommended: rephrasing

> 
> The purpose of this document is to create a Double-Signature SBT standard by creating consensus around a maximally [EIP-1155](./eip-1155.md) compatible interface definition.

Reviewer: ```suggestion
```
Recommended: Implicit or previously stated

### Characteristics

> - The NFT will be non-transferable after the initial transfer(Soulbound/Accountbound Token-SBT).

Reviewer: ```suggestion
- The NFT will be non-transferable after the initial transfer
```

        uint256 id,
        bool action
    ) external;

    /**
     * @dev Claims or Reject pending tokens under `ids[]`.
     *
     * Requirements for each `id` `action` pair:
     * - `account` must have a pending token under `id` at the moment of call.
     * - `account` must not own a token under `id` at the moment of call.
     *
     * Emits a {TokenClaimed} event.
     *
     */
    function claimOrRejectBatch(
        address account,
        uint256[] memory ids,
        bool[] memory actions
    ) external;

    /**
     * @dev Transfers `_id` token from `_from` to every address at `_to[]`.
     *
     * Requirements:
     *
     * - `_from` MUST be the creator(minter) of `id`.
     * - All addresses in `to[]` MUST be non-zero.
     * - All addresses in `to[]` MUST have the token `id` under `_pendings`.
     * - All addresses in `to[]` MUST not own a token type under `id`.
     *
     * Emits a {TransfersMulti} event.
     *
     */
    function batchTransfer(
        address from,
        address[] memory to,
        uint256 id,
        uint256 amount,
        bytes memory data
    ) external;
    
}

```

## Rationale

### SBT as an extension to EIP-1155

> We believe that Soulbound Tokens serve as a specialized subset of the existing [EIP-1155](./eip-1155.md) tokens. The advantage of such a design is the seamless compatibility of SBTs with existing NFT services. Service providers can treat SBTs like NFTs and do not need to make drastic changes to their existing codebase.

Author: Changed backwards compatibility to partial compatibility.

> We believe that Soulbound Token serves as a specialized subset of the existing [EIP-1155](./eip-1155.md) tokens. The advantage of such design is the seamless compatibility of Soulbound Token with existing NFT services. Service providers can treat SBTs like NFTs and do not need to make drastic changes to their existing codebase.

Reviewer: ```suggestion
We believe that Soulbound Tokens serve as a specialized subset of the existing [EIP-1155](./eip-1155.md) tokens. The advantage of such a design is the seamless compatibility of Soulbound Token with existing NFT services. Service providers can treat SBTs like NFTs and do not need to make drastic changes to their existing codebase.
```

> Making the standard compatible with [EIP-1155](./eip-1155.md) also allows for SBTs to bind to multiple addresses and to Smart Contracts.

Reviewer: ```suggestion
Making the standard mostly compatible with [EIP-1155](./eip-1155.md) also allows for SBTs to bind to multiple addresses and to Smart Contracts.
```
Required: This is not compatible with EIP-1155

> Making the standard compatible with [EIP-1155](./eip-1155.md) also allows for SBT binding to Smart Contracts,allowing for a whole new world of possibilities.

Reviewer: ```suggestion
Making the standard compatible with [EIP-1155](./eip-1155.md) also allows for SBT binding to Smart Contracts, allowing for a whole new world of possibilities.
```

### Metadata.
The [EIP-1155](./eip-1155.md#metadata) Metadata Interface was implemented for more compatibility with [EIP-1155](./eip-1155.md).

### Guaranteed log trace
> As the Ethereum ecosystem continues to grow, many DApps are relying on traditional databases and explorer API services to retrieve and categorize data. The EIP-1155 standard guarantees that event logs emitted by the smart contract will provide enough data to create an accurate record of all current token balances. A database or explorer may listen to events and be able to provide indexed and categorized searches of every EIP-1155 token in the contract.

_quoted from [EIP-1155](./eip-1155.md#guaranteed-log-trace)_

This EIP extends this idea to the Double Signature functionality: The `{TokenClaimed}` event logs all the necessary information of a `ClaimOrReject(...)` or `ClaimOrRejectBatch(...)` function call, storing relevant information about the actions performed by the user. This also applies to the `batchTransfer(...)` function: It emits the `{TransferMulti}` event and logs necessary data.

### Exception handling

> Given the non-transferability property of SBTs, if a user keys to an account or a contract get compromised or rotated, a user may lose the ability to associate themselves with the token.

Reviewer: ```suggestion
Given the non-transferability property of SBTs, if a user's keys to an account or a contract get compromised or rotated, a user may lose the ability to associate themselves with the token.
```

**Given the multi-owner characteristic of [EIP-1155](./eip-1155.md) compliant interfaces and contracts, this allows for SBTs to bind to multiple accounts, providing a potential solution to the issue.**

> Multi-owned SBTs can also be issued to a contract account that implements a multi-signature functionality (As recommended in [EIP-4973](./eip-4973.md#exception-handling)), this can be achieved via the [EIP-1155](./eip-1155.md#erc-1155-token-receiver) Token Receiver interface.

Reviewer: ```suggestion
Multi-owned SBTs can also be issued to a contract account that implements a multi-signature functionality (As recommended in [EIP-4973](./eip-4973.md#exception-handling)); this can be achieved via the [EIP-1155](./eip-1155.md#erc-1155-token-receiver) Token Receiver interface.
```

### Multi-token
The multi-token functionality permits the implementation of multiple token types in the same contract. Furthermore, all emitted tokens are stored in the same contract, preventing redundant bytecode from being deployed to the blockchain. It also facilitates transfer to token issuers, since all issued tokens are stored and can be accessed under the same contract address.

> ### batchTransfer(...) function

Reviewer: ```suggestion
### The `batchTransfer` function
```
Recommended: formatting

> The multi-token functionality was used to ease organization. All emitted tokens are stored in the same contract, preventing redundant bytecode from being deployed to the blockchain. It also facilitates transfer to token issuers, since all issued tokens are stored and can be accessed under the same address. It is also easy to implement non-fungible and semi-fungible tokens in the same contract.

Reviewer: ```suggestion
The multi-token functionality was used to ease organization. All emitted tokens are stored in the same contract, preventing redundant bytecode from being deployed to the blockchain. It also facilitates transfer to token issuers, since all issued tokens are stored and can be accessed under the same address. It is also easy to implement non-fungible and semi-fungible tokens in the same contract.
```
Defile "ease organization.

This EIP support transfers to multiple recipients. This eases transfer to a large number of people, making it more gas-efficient and user-friendly.

## Backwards Compatibility
This proposal is fully backward compatible with EIP-1155.

## Reference Implementation
You can find an implementation of this standard in [../assets/EIP-5516](../assets/eip-5516/ERC5516.sol).

## Security Considerations

> There are no security considerations pertinent to the implementation of this standard.

Reviewer: ```suggestion
Needs discussion.
```

---
eip: 5516

> title: Soulbound, Multi-Token Standard.

Reviewer: ```suggestion
title: Soulbound Multi-Tokens
```

> description: An interface for non-transferrable NFTs binding to an Ethereum account for standardized blockchain-based academic certification.

Reviewer: ```suggestion
description: An interface for non-transferrable NFTs binding to an Ethereum account for blockchain-based certification
```

## Abstract

This is a standard interface for non fungible, double signature, multi-token, Soulbound Tokens. It was inspired by the main characteristics of the EIP-1155 token and by articles in which benefits and potential use-cases of Soulbound/Accountbound Tokens(SBTs) were presented.

## Motivation

> Current Soulbound/Accountbound token standards like EIP-4973 face the potential issue of the user losing their account keys or having them rotated, therefore losing their bound to the SBTs. This EIP-1555 inspired interface design provides a solution to this issue and also allows for a recycling of SBTs: minting of NFTs with the same characteristics for it being transferred to different addresses is no longer neccesary.

Reviewer: ```suggestion
Current Soulbound/Accountbound token standards like EIP-4973 face the potential issue of the user losing their account keys or having them rotated, therefore losing their bound to the SBTs. This [EIP-1555](./eip-1155.md)-inspired interface design provides a solution to this issue and also allows for recycling of SBTs: minting of NFTs with the same characteristics for it being transferred to different addresses is no longer necessary.
```

     * Emits a {TransfersMulti} event.
     *
     */
    function batchTransfer (address from, address[] memory to, uint256 id, uint256 amount, bytes memory data) external;

}
```

## Rationale

### SBT as an extension to EIP-1155
We believe that Soulbound Token serves as a specialized subset of the existing EIP-1155 tokens. The advantage of such design is seamless compatibility of Soulbound Token with existing NFT services. Service providers can treat SBTs like NFTs and do not need to make drastic changes to their existing codebase.
Making the standard compatible with EIP-1155 also allows for SBT binding to Smart Contracts, and this can allow for a whole new world of possibilities.

### Double-Signature
The Double-Signature functionality was implemented to prevent the receival of unwanted tokens. It symbolizes a handshake between token receiver and token sender, implying that **both** parties agree on the token transfer.

### Metadata.
The EIP-1155 Metadata Interface was implemented for more compatibility with EIP-1155.   

_See [EIP-1155](./EIP-1155.md#metadata)_ 

### Guaranteed log trace

> >As the Ethereum ecosystem continues to grow, many dapps are relying on traditional databases and explorer API services to retrieve and categorize data. The ERC-1155 standard guarantees that event logs emitted by the smart contract will provide enough data to create an accurate record of all current token balances. A database or explorer may listen to events and be able to provide indexed and categorized searches of every ERC-1155 token in the contract.

Reviewer: ```suggestion
> As the Ethereum ecosystem continues to grow, many DApps are relying on traditional databases and explorer API services to retrieve and categorize data. The EIP-1155 standard guarantees that event logs emitted by the smart contract will provide enough data to create an accurate record of all current token balances. A database or explorer may listen to events and be able to provide indexed and categorized searches of every EIP-1155 token in the contract.
```
